lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.cpp

```diff
@@ -0,0 +1,708 @@
+
+// Type traits for GfVec precision.
+template<class V>
+struct OpPrecision {
+    static UsdGeomXformOp::Precision precision;
+};
+
+template<>
+UsdGeomXformOp::Precision OpPrecision<GfVec3f>::precision = 
+    UsdGeomXformOp::PrecisionFloat;
+
+template<>
+UsdGeomXformOp::Precision OpPrecision<GfVec3d>::precision = 
+    UsdGeomXformOp::PrecisionDouble;
+
+inline double TO_DEG(double a) { return a * 180.0 / 3.141592654; }
+inline double TO_RAD(double a) { return a * 3.141592654 / 180.0; }
+
+VtValue getValue(const UsdAttribute& attr, const UsdTimeCode& time)
+{
+    VtValue value;
+    attr.Get(&value, time);
+    return value;
+}
+
+// UsdMayaXformStack::FindOpIndex() requires an inconvenient isInvertedTwin
+// argument, various rotate transform op equivalences in a separate
+// UsdMayaXformStack::IsCompatibleType().  Just roll our own op name to
+// Maya transform stack index position.
+const std::unordered_map<std::string, UsdTransform3dMayaXformStack::OpNdx> opNameToNdx{
+    {"xformOp:translate",           UsdTransform3dMayaXformStack::NdxTranslate},
+    {"xformOp:translate:rotatePivotTranslate", UsdTransform3dMayaXformStack::NdxRotatePivotTranslate},
+    {"xformOp:translate:rotatePivot", UsdTransform3dMayaXformStack::NdxRotatePivot},
+    {"xformOp:rotateX",             UsdTransform3dMayaXformStack::NdxRotate},
+    {"xformOp:rotateY",             UsdTransform3dMayaXformStack::NdxRotate},
+    {"xformOp:rotateZ",             UsdTransform3dMayaXformStack::NdxRotate},
+    {"xformOp:rotateXYZ",           UsdTransform3dMayaXformStack::NdxRotate},
+    {"xformOp:rotateXZY",           UsdTransform3dMayaXformStack::NdxRotate},
+    {"xformOp:rotateYXZ",           UsdTransform3dMayaXformStack::NdxRotate},
+    {"xformOp:rotateYZX",           UsdTransform3dMayaXformStack::NdxRotate},
+    {"xformOp:rotateZXY",           UsdTransform3dMayaXformStack::NdxRotate},
+    {"xformOp:rotateZYX",           UsdTransform3dMayaXformStack::NdxRotate},
+    {"xformOp:orient",              UsdTransform3dMayaXformStack::NdxRotate},
+    {"xformOp:rotateXYZ:rotateAxis", UsdTransform3dMayaXformStack::NdxRotateAxis},
+    {"!invert!xformOp:translate:rotatePivot", UsdTransform3dMayaXformStack::NdxRotatePivotInverse},
+    {"xformOp:translate:scalePivotTranslate", UsdTransform3dMayaXformStack::NdxScalePivotTranslate},
+    {"xformOp:translate:scalePivot", UsdTransform3dMayaXformStack::NdxScalePivot},
+    {"xformOp:shear",               UsdTransform3dMayaXformStack::NdxShear},
+    {"xformOp:scale",               UsdTransform3dMayaXformStack::NdxScale},
+    {"!invert!xformOp:translate:scalePivot", UsdTransform3dMayaXformStack::NdxScalePivotInverse}};
+
+//----------------------------------------------------------------------
+// Conversion functions from RotXYZ to all supported rotation attributes.
+//----------------------------------------------------------------------
+
+typedef VtValue (*CvtRotXYZToAttrFn)(double x, double y, double z);
```

Comment:
UFE requires RotXYZ Euler angle rotation ordering, so when writing to USD attribute use the appropriate conversion function.